go/vt/schemadiff/errors.go

```diff
@@ -301,6 +301,44 @@ func (e *ForeignKeyColumnTypeMismatchError) Error() string {
 	)
 }
 
+type MissingForeignKeyIndexError struct {
+	Table      string
+	Constraint string
+}
+
+func (e *MissingForeignKeyIndexError) Error() string {
+	return fmt.Sprintf("missing index for foreign key constraint %s in table %s",
+		sqlescape.EscapeID(e.Constraint),
+		sqlescape.EscapeID(e.Table),
+	)
+}
+
+type MissingForeignKeyReferencedIndexError struct {
+	Table           string
+	Constraint      string
+	ReferencedTable string
+}
+
+func (e *MissingForeignKeyReferencedIndexError) Error() string {
+	return fmt.Sprintf("missing index in referenced table %s for foreign key constraint %s in table %s",
+		sqlescape.EscapeID(e.ReferencedTable),
+		sqlescape.EscapeID(e.Constraint),
+		sqlescape.EscapeID(e.Table),
+	)
+}
+
+type IndexNeededByForeignKeyError struct {
+	Table string
+	Key   string
+}
+
+func (e *IndexNeededByForeignKeyError) Error() string {
+	return fmt.Sprintf("key %s needed by a foreign key constraint in table %s",
+		sqlescape.EscapeID(e.Key),
+		sqlescape.EscapeID(e.Table),
+	)
+}
```

Comment:
Good catch on the 1st error, `MissingForeignKeyIndexError`: I've just **removed it**. It can never fire because, compatible with MySQL's behavior, we always add an index if one is needed, either in a `CREATE TABLE` statement or in a `ALTER TABLE ... ADD ... FOREIGN KEY` statement.

2nd error, `MissingForeignKeyReferencedIndexError`, is when the _parent_ table does not have an index covering the referenced columns. Here, MySQL does not add anything implicitly, because that's on a different table. The error is compatible with MySQL.

3rd error, `IndexNeededByForeignKeyError` is for when you attempt to `ALTER TABLE ... DROP KEY` and when that leaves a foreign key without an index covering its local columns. Compatible with MySQL, that's an error.